Firebend.AutoCrud.Core/Implementations/Entities/ClientRequestTransactionManager.cs

```diff
@@ -8,65 +8,70 @@
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
 
 namespace Firebend.AutoCrud.Core.Implementations.Entities;
 
 public class ClientRequestTransactionManager : ISessionTransactionManager
 {
-    private readonly IServiceScopeFactory _scopeFactory;
+    private class QueuedTransaction
+    {
+        public IEntityTransaction Transaction { get; init; }
+        public DateTimeOffset StartedDate { get; init; }
+        public bool Removed { get; set; }
+    }
 
-    private readonly ConcurrentDictionary<string, Task<IEntityTransaction>> _sharedTransactions = new();
-    private readonly ConcurrentQueue<IEntityTransaction> _transactions = new();
+    private readonly string _sessionId = Guid.NewGuid().ToString();
+    private readonly ILogger<ClientRequestTransactionManager> _logger;
+    private readonly IServiceProvider _serviceProvider;
+
+    private readonly ConcurrentDictionary<string, Lazy<Task<IEntityTransaction>>> _sharedTransactions = new();
+    private readonly ConcurrentBag<QueuedTransaction> _transactions = new();
 
     public bool TransactionStarted { get; private set; }
-    public ImmutableList<Guid> TransactionIds => _transactions.Select(x => x.Id).ToImmutableList();
+    public ImmutableList<Guid> TransactionIds => GetTransactionsInOrder(true).Select(x => x.Id).ToImmutableList();
 
-    public ClientRequestTransactionManager(IServiceScopeFactory scopeFactory)
+    public ClientRequestTransactionManager(ILogger<ClientRequestTransactionManager> logger,
+        IServiceProvider serviceProvider)
     {
-        _scopeFactory = scopeFactory;
+        _logger = logger;
+        _serviceProvider = serviceProvider;
     }
 
-    public void Start() => TransactionStarted = true;
+    public void Start()
+    {
+        _logger.LogDebug("{SessionId}: Starting session transaction", _sessionId);
+        TransactionStarted = true;
+    }
 
     private IEnumerable<IEntityTransaction> GetTransactionsInOrder(bool ascending)
     {
-        IEnumerable<IEntityTransaction> EntityTransactions()
-        {
-            while (_transactions.TryDequeue(out var t))
-            {
-                yield return t;
-            }
-        }
-
-        if (ascending)
-        {
-            return EntityTransactions().OrderBy(x => x.StartedDate).ToArray();
-        }
-
-        return EntityTransactions().OrderByDescending(x => x.StartedDate).ToArray();
+        var activeTransactions = _transactions.Where(x => !x.Removed);
+        var queuedTransactionsOrdered = ascending
+            ? activeTransactions.OrderBy(x => x.StartedDate)
+            : activeTransactions.OrderByDescending(x => x.StartedDate);
+        return queuedTransactionsOrdered.Select(x => x.Transaction);
     }
 
     public async Task CompleteAsync(CancellationToken cancellationToken)
     {
+        _logger.LogDebug("{SessionId}: Completing {TransactionCount} transactions. {TransactionIds}", _sessionId,
```

Comment:
we should find a different way to show all the transaction ids in this log statement. regardless of the log level `string.join` will get called allocating memory. 

we could wrap the statement in a `if (_logger.IsEnabled(LogLevel.Debug))` block